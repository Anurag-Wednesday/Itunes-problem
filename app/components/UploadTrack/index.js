import React, { memo, useState } from 'react';
import { Steps, Button, Form, Input, DatePicker, Upload } from 'antd';
import { UploadOutlined } from '@ant-design/icons';
import { FormattedMessage } from 'react-intl';
import styled from 'styled-components';
import * as colors from '@app/themes/colors';
import { TrackDetailsComponent } from '../TrackDetailsComponent/index';

const StyledCard = styled.div`
  && {
    background-color: ${colors.background};
    padding: 5rem;
  }
`;

const StyledInput = styled(Input)`
  && {
    width: 25rem;
  }
`;

const StyledFormItem = styled(Form.Item)`
  && {
    margin: 2rem;
    color: ${colors.whiteText};
    wrapper
  }
  .ant-form-item-required {
    color: ${colors.whiteText};
    font-size: 1rem;
  }
`;

const TrackContainer = styled.div`
  && {
    display: flex;
    align-items: center;
    justify-content: center;
    background-color: ${colors.background};
    min-width: 100vw;
    border: 1px solid ${colors.primary};
    min-height: 100vh;
  }
`;

const { Step } = Steps;

export function UploadTrack() {
  const [data, setData] = useState({});
  const [current, setCurrent] = useState(0);
  const onFinish = (values) => {
    setData(Object.assign(data, values));
    setCurrent(current + 1);
  };

  const prev = () => {
    setCurrent(current - 1);
  };

  const contentOne = () => {
    return (
      <>
        <Form
          data-testid="form"
          name="basic"
          labelCol={{ span: 8 }}
          wrapperCol={{ span: 16 }}
          initialValues={{ remember: false }}
          autoComplete="on"
          onFinish={onFinish}
        >
          <StyledFormItem
            label={<FormattedMessage id="input-track" />}
            name="trackName"
            rules={[{ required: true, message: <FormattedMessage id="input-track" /> }]}
          >
            <StyledInput data-testid="track-input" />
          </StyledFormItem>

          <StyledFormItem
            label={<FormattedMessage id="input-artist" />}
            name="artistName"
            rules={[{ required: true, message: <FormattedMessage id="input-artist" /> }]}
          >
            <StyledInput data-testid="artist-name" />
          </StyledFormItem>
          <StyledFormItem
            label={<FormattedMessage id="input-collection" />}
            name="collectionName"
            rules={[{ required: true, message: <FormattedMessage id="input-collection" /> }]}
          >
            <StyledInput data-testid="collection-name" />
          </StyledFormItem>
          <StyledFormItem wrapperCol={{ offset: 8, span: 16 }}>
            <Button type="primary" htmlType="submit" data-testid="submit">
              Next
            </Button>
          </StyledFormItem>
        </Form>
      </>
    );
  };

  const contentTwo = () => {
    return (
      <>
        <Form
          name="basic"
          labelCol={{ span: 8 }}
          wrapperCol={{ span: 16 }}
          initialValues={{ remember: false }}
          autoComplete="on"
          onFinish={onFinish}
        >
          <StyledFormItem
            label={<FormattedMessage id="input-genre" />}
            name="primaryGenreName"
            rules={[{ required: true, message: <FormattedMessage id="input-genre" /> }]}
          >
            <StyledInput data-testid="genre-input" />
          </StyledFormItem>
          <StyledFormItem
            label={<FormattedMessage id="input-genre" />}
            name="country"
            rules={[{ required: true, message: <FormattedMessage id="input-country" /> }]}
          >
            <StyledInput data-testid="country-input" />
          </StyledFormItem>
          <StyledFormItem
            label={<FormattedMessage id="input-date" />}
            name="releaseDate"
            rules={[{ required: false, message: <FormattedMessage id="input-date" /> }]}
          >
            <DatePicker />
          </StyledFormItem>
          <StyledFormItem wrapperCol={{ offset: 8, span: 16 }}>
            <Button type="primary" htmlType="submit" data-testid="submit">
              Next
            </Button>
          </StyledFormItem>
        </Form>
      </>
    );
  };

  const contentThree = () => {
    return (
      <>
        <Form
          name="basic"
          labelCol={{ span: 8 }}
          wrapperCol={{ span: 16 }}
          initialValues={{ remember: false }}
          autoComplete="on"
          onFinish={onFinish}
        >
          <StyledFormItem
            label={<FormattedMessage id="input-date" />}
            name="artWorkUrl100"
            rules={[{ required: true, message: <FormattedMessage id="input-date" /> }]}
          >
            <Upload data-testid="upload">
              <Button icon={<UploadOutlined />}>Click to Upload</Button>
            </Upload>
          </StyledFormItem>
          <StyledFormItem wrapperCol={{ offset: 8, span: 16 }}>
            <Button type="primary" htmlType="submit" data-testid="submit">
              Submit
            </Button>
          </StyledFormItem>
        </Form>
      </>
    );
  };
  const content = [contentOne, contentTwo, contentThree];

  return (
    <>
      <StyledCard data-testid="upload-track">
        <Steps current={current}>
          <Step />
          <Step />
          <Step />
        </Steps>
        <div className="steps-content">{content[current]()}</div>
        <div className="steps-action">
          {current > 0 && (
            <Button data-testid="previous" style={{ margin: '0 8px' }} onClick={() => prev()}>
              Previous
            </Button>
          )}
        </div>
      </StyledCard>
      <TrackContainer>
        <TrackDetailsComponent {...data} />
      </TrackContainer>
    </>
  );
}

UploadTrack.propTypes = {};

export default memo(UploadTrack);
