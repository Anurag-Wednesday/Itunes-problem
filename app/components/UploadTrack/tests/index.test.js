/**
 *
 * Tests for UploadTrack
 *
 */

import React from 'react';
import { fireEvent } from '@testing-library/dom';
import { renderWithIntl, timeout } from '@utils/testUtils';
import UploadTrack from '../index';

describe('<UploadTrack />', () => {
  it('should render and match the snapshot', () => {
    const { baseElement } = renderWithIntl(<UploadTrack />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should contain 1 UploadTrack component', () => {
    const { getAllByTestId } = renderWithIntl(<UploadTrack />);
    expect(getAllByTestId('upload-track').length).toBe(1);
  });

  it('should call the onFinish with values in the input', async () => {
    const { getByTestId } = renderWithIntl(<UploadTrack />);
    const trackName = 'Sad';
    const artistName = 'Sad';
    const collection = 'Sad';
    fireEvent.change(getByTestId('track-input'), { target: { value: trackName } });
    fireEvent.change(getByTestId('artist-name'), { target: { value: artistName } });
    fireEvent.change(getByTestId('collection-name'), { target: { value: collection } });
    fireEvent.click(getByTestId('submit'));
    await timeout(500);
    expect(getByTestId('track_title')).toHaveTextContent(trackName);
  });
  it('should take the user back to the previous test if clicked on previous button', async () => {
    const { getByTestId } = renderWithIntl(<UploadTrack />);
    const trackName = 'Sad';
    const artistName = 'Sad';
    const collection = 'Sad';
    fireEvent.change(getByTestId('track-input'), { target: { value: trackName } });
    fireEvent.change(getByTestId('artist-name'), { target: { value: artistName } });
    fireEvent.change(getByTestId('collection-name'), { target: { value: collection } });
    fireEvent.click(getByTestId('submit'));
    await timeout(500);
    fireEvent.click(getByTestId('previous'));
    expect(getByTestId('track-input')).toBeInTheDocument();
  });
  it('should give the user option tom upload image and submit once on step 3', async () => {
    const { getByTestId } = renderWithIntl(<UploadTrack />);
    const trackName = 'Sad';
    const artistName = 'Sad';
    const collection = 'Sad';
    fireEvent.change(getByTestId('track-input'), { target: { value: trackName } });
    fireEvent.change(getByTestId('artist-name'), { target: { value: artistName } });
    fireEvent.change(getByTestId('collection-name'), { target: { value: collection } });
    fireEvent.click(getByTestId('submit'));
    await timeout(500);
    fireEvent.change(getByTestId('genre-input'), { target: { value: artistName } });
    fireEvent.change(getByTestId('country-input'), { target: { value: collection } });
    fireEvent.click(getByTestId('submit'));
    await timeout(500);
    expect(getByTestId('upload')).toBeInTheDocument();
    expect(getByTestId('submit')).toBeInTheDocument();
  });
});
