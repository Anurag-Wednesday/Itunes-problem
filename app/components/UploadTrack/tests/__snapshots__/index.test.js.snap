// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UploadTrack /> should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="UploadTrack__StyledCard-sc-7ttrjy-0 eNWSuX"
      data-testid="upload-track"
    >
      <div
        class="ant-steps ant-steps-horizontal ant-steps-label-horizontal"
      >
        <div
          class="ant-steps-item ant-steps-item-process ant-steps-item-active"
        >
          <div
            class="ant-steps-item-container"
          >
            <div
              class="ant-steps-item-tail"
            />
            <div
              class="ant-steps-item-icon"
            >
              <span
                class="ant-steps-icon"
              >
                1
              </span>
            </div>
            <div
              class="ant-steps-item-content"
            >
              <div
                class="ant-steps-item-title"
              />
            </div>
          </div>
        </div>
        <div
          class="ant-steps-item ant-steps-item-wait"
        >
          <div
            class="ant-steps-item-container"
          >
            <div
              class="ant-steps-item-tail"
            />
            <div
              class="ant-steps-item-icon"
            >
              <span
                class="ant-steps-icon"
              >
                2
              </span>
            </div>
            <div
              class="ant-steps-item-content"
            >
              <div
                class="ant-steps-item-title"
              />
            </div>
          </div>
        </div>
        <div
          class="ant-steps-item ant-steps-item-wait"
        >
          <div
            class="ant-steps-item-container"
          >
            <div
              class="ant-steps-item-tail"
            />
            <div
              class="ant-steps-item-icon"
            >
              <span
                class="ant-steps-icon"
              >
                3
              </span>
            </div>
            <div
              class="ant-steps-item-content"
            >
              <div
                class="ant-steps-item-title"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="steps-content"
      >
        <form
          autocomplete="on"
          class="ant-form ant-form-horizontal"
          data-testid="form"
          id="basic"
        >
          <div
            class="ant-row ant-form-item UploadTrack__StyledFormItem-sc-7ttrjy-2 xcKPX"
          >
            <div
              class="ant-col ant-col-8 ant-form-item-label"
            >
              <label
                class="ant-form-item-required"
                for="basic_trackName"

                title=""
              >
                Please enter Track Name
              </label>
            </div>
            <div
              class="ant-col ant-col-16 ant-form-item-control"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <input
                    class="ant-input UploadTrack__StyledInput-sc-7ttrjy-1 htJzPi"
                    data-testid="track-input"
                    id="basic_trackName"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="ant-row ant-form-item UploadTrack__StyledFormItem-sc-7ttrjy-2 xcKPX"
          >
            <div
              class="ant-col ant-col-8 ant-form-item-label"
            >
              <label
                class="ant-form-item-required"
                for="basic_artistName"
                title=""
              >
                Please enter Artist Name

              </label>
            </div>
            <div
              class="ant-col ant-col-16 ant-form-item-control"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <input
                    class="ant-input UploadTrack__StyledInput-sc-7ttrjy-1 htJzPi"
                    data-testid="artist-name"
                    id="basic_artistName"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="ant-row ant-form-item UploadTrack__StyledFormItem-sc-7ttrjy-2 xcKPX"
          >
            <div
              class="ant-col ant-col-8 ant-form-item-label"
            >
              <label
                class="ant-form-item-required"
                for="basic_collectionName"

                title=""
              >
                Please enter Collection Name
              </label>
            </div>
            <div
              class="ant-col ant-col-16 ant-form-item-control"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <input
                    class="ant-input UploadTrack__StyledInput-sc-7ttrjy-1 htJzPi"
                    data-testid="collection-name"
                    id="basic_collectionName"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="ant-row ant-form-item UploadTrack__StyledFormItem-sc-7ttrjy-2 xcKPX"
          >
            <div
              class="ant-col ant-col-16 ant-col-offset-8 ant-form-item-control"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <button
                    class="ant-btn ant-btn-primary"
                    data-testid="submit"
                    type="submit"
                  >
                    <span>
                      Next
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </form>
      </div>
      <div
        class="steps-action"
      />
    </div>
    <div
      class="UploadTrack__TrackContainer-sc-7ttrjy-3 kDiFcG"
    >
      <div
        class="TrackDetailsComponent__TrackDetailsCard-sc-1q5q059-7 ijfgSK"
        data-testid="track-details-component"
      >
        <p
          class="T__StyledText-gjlic1-0 iVJYgn"
          data-testid="image_unavailable"
        >
          Image Unavailable
        </p>
        <div
          class="TrackDetailsComponent__Details-sc-1q5q059-0 ecyiHy"
        >
          <div
            class="TrackDetailsComponent__TrackTitle-sc-1q5q059-2 aNhbq"
          >
            <p
              class="T__StyledText-gjlic1-0 iVJYgn"
              data-testid="no-track-name"
            >
              No track name available
            </p>
          </div>
          <div
            class="TrackDetailsComponent__ArtistName-sc-1q5q059-3 jtjBVl"
          >
            <p
              class="T__StyledText-gjlic1-0 iVJYgn"
              data-testid="t"
            >
              No artist data available
            </p>
          </div>
          <p
            class="T__StyledText-gjlic1-0 iVJYgn"
            data-testid="artist_data_unavailable"
          >
            No artist data available
          </p>
          <p
            class="T__StyledText-gjlic1-0 iVJYgn"
            data-testid="t"
          >
            No artist data available
          </p>
          <p
            class="T__StyledText-gjlic1-0 iVJYgn"
            data-testid="t"
          >
            No preview available
          </p>
          <p
            class="T__StyledText-gjlic1-0 iVJYgn"
            data-testid="t"
          >
            No artist data available
          </p>
        </div>
      </div>
    </div>
  </div>
</body>
`;
